src/Data/Form.purs

```diff
@@ -41,6 +43,7 @@ type Field f = Record (FieldRows f ())
 type SharedRows f r =
   ( default :: Maybe f
   , value :: InputSource f
+  , errors :: Errors ValidationError
```

Comment:
@joneshf-cn thanks for pointing these issues out, you're 💯 right. So maybe we merge `ValidationError` in with `InputSource`? The question then is how?
```purescript
data InputSource e a
  = NotSet
  | UserCleared
  | Invalid e a
  | UserSet a
```
Still doesn't seem quite right, as `NotSet` and `UserCleared` could be invalid, depending on whether the field is `required`.

Plus there's the `default` value to think about. I don't know that we can roll validation into `InputSource` without rolling `default` value in as well. Plus `InputSource` is not actually meant to act as validation, but as a means of knowing how the value was set. It was either not set, cleared by the user, cleared by form logic (my original intent was that invalid dropdown values would be cleared, but how I implemented it, they aren't, we just show an error), or set by the user.

Adding in validation changes the meaning and potentially usefulness of `InputSource`. Maybe the solution is separating `InputSource` from any value.

```purescript
data InputSource
  = Form
  | User

data ValidationResult a
  = Valid a
  | Required
  | Invalid a
  | MinLength Int a
  | MaxLength Int a
```
This covers the `NotSet` and `UserCleared` cases, because if they aren't required our `a` would be a monoid (or at least a `Maybe`).

Then we would have an unvalidated version of our `Input` that only has a value, and a validated one.

```purescript
toggleExprType :: Input ExprType
toggleExprType = Toggle
  { default: Just (Boolean true) :: Maybe ExprType
  , required: Just (Boolean true) :: Maybe ExprType
  , source: Form :: InputSource
  , value: Nothing :: Maybe ExprType
  }

toggleValidated :: Input Validated
toggleValidated = Toggle
  { default: Just (Boolean true) :: Maybe ExprType
  , required: Just (Boolean true) :: Maybe ExprType
  , source: Form :: InputSource
  , value: Valid (Boolean true) :: ValidationResult (Maybe ExprType)
  }
```
Though at that point, if we can, maybe we should just get rid of the `default` and `source` labels from our `Input Validated` record, as it's not useful anymore and `default` can yield invalid state.

There are a couple issues. This doesn't allow for multiple validation errors. Though I don't know yet that we need them. `[Required, MinLength 1]` is redundant.

Also, if we adopted this route we'd prob want to make `ValidationResult` a `Variant` instead of sum type, so that we can't have `MinLength 2` on a `Toggle`.

Reply:
> `NotSet` and `UserCleared` could be invalid, depending on whether the field is `required`.

This line I think says everything we need to know. `required` isn't a `Boolean` that we product onto `InputSource _`. `required` determines what `InputSource _` is. We don't have:

```PureScript
data Input a
  = Currency
      { required :: Boolean
      , value :: InputSource a
      , …
      }
  | Dropdown
      { required :: Boolean
      , value :: InputSource a
      , …
      }

data InputSource a
  = UserInput a
  | Invalid a
  | UserCleared
  | NotSet
```

We have something more like:
```PureScript
data Input a
  = OptionalCurrency
      { value :: OptionalValue a
      , …
      }
  | OptionalDropdown
      { value :: OptionalValue a
      , …
      }
  | RequiredCurrency
      { value :: RequiredValue a
      , …
      }
  | RequiredDropdown
      { value :: RequiredValue a
      , …
      }

data OptionalValue a
  = OptionalMin Int a
  | OptionalMax Int a
  | OptionalInput a

data RequiredValue a
  = NotSet
  | UserCleared
  | RequiredMin Int a
  | RequiredMax Int a
  | RequiredInput a
```
Or whatever.

We can manipulate the data types to be more ergonomic, but what's important is that the former encoding with `required :: Boolean` had `4 + 4a` states for just `Currency`s value while the suggested one has `2 + 6a` states for just `Currency`s value.

Reply:
That's not to say we have to decode/encode directly to that, we can keep the serialization simpler for our consumers, but we should work with a representation that more correctly matches our mental model of what the form is.

Reply:
I felt like this needed a bit of rethinking, so I went over to Notion for that: https://www.notion.so/citizennet/Rethinking-Input-Types-6b87407f5872419ca50730d523745eee

Let me know your thoughts @joneshf-cn

Reply:
I'm sure you've seen, but I left some comments in the notion. Ultimately, I think we should decide on something and move on. Having any sort of validation is better than none, and I don't want this PR to stymie more while trying to find the perfect representation. If you want to explore what you've written, I say let's go for it!

Reply:
@joneshf-cn how do you feel about merging this in and improving it in a separate PR? It's already getting so big that any more changes are going to be difficult to parse out. And it _does_ work. I've merged in `master` and resolved conflicts, so it's good to go.

I do want to address the issues brought up, but the ideas I've drawn up in Notion aren't thought out enough. E.g., turns out we do need something like `NotSet`, so that we can know whether to display validation errors or not. It's possible that can be accomplished another way or with another value, but I think these types still need more thought.

Reply:
I'd also be okay with not merging it in until these issues are addressed. Will just have to periodically keep it up to date with `master`, but that would at least force us to try to answer these issues sooner than later

Reply:
Merge it up!